### PR TITLE
PREAPPS-2439: 2FA | Enable and Setup 2FA

### PR DESCRIPTION
**Pull request by panditsavitags**

### CODE REVIEW

File: src/batch-client/index.ts [hunk 1]
```diff
@@ -432,6 +433,28 @@ export class ZimbraBatchClient {
...
+				name: {
+					_content: options.name
+				},
+				password: options.password
+					? {
+							_content: options.password
+					  }
+					: null,
+				authToken: options.authToken,
+				twoFactorCode: options.twoFactorCode
```
Comment (silentsakky) [inline src/batch-client/index.ts]: use destructuring to add optional elements in object

... twoFactorCode && {
     twoFactorCode
}

File: src/batch-client/index.ts [hunk 2]
```diff
@@ -432,6 +433,28 @@ export class ZimbraBatchClient {
 			content
 		}));
 
+	public enableTwoFactorAuth = (options: EnableTwoFactorAuthInput) =>
+		this.jsonRequest({
+			name: 'EnableTwoFactorAuth',
+			body: {
+				name: {
+					_content: options.name
```
Comment (silentsakky) [inline src/batch-client/index.ts]: destructure all options in function argument itself

File: src/batch-client/index.ts [hunk 3]
```diff
@@ -432,6 +433,28 @@ export class ZimbraBatchClient {
...
+				},
+				password: options.password
+					? {
+							_content: options.password
```
Comment (silentsakky) [inline src/batch-client/index.ts]: not sure why we need to send everything in _content property
you can directly send it like `password: options.password`

File: src/schema/schema.graphql [hunk 4]
```diff
@@ -2623,6 +2657,7 @@ type Mutation {
 		twoFactorCode: String,
 		deviceTrusted: Boolean
 	): AuthResponse
+	enableTwoFactorAuth( options: EnableTwoFactorAuthInput! ) : EnableTwoFactorAuthResponse
```
Comment (silentsakky) [inline src/schema/schema.graphql]: remove extra spaces from function arguments

File: src/schema/schema.graphql [hunk 5]
```diff
@@ -2237,6 +2265,12 @@ input ExternalAccountModifyAttrsInput {
 	port: String
 	smtpPort: String
 }
+input EnableTwoFactorAuthInput {
+	authToken: [TwoFactorAuthTokenInput]
```
Comment (silentsakky) [inline src/schema/schema.graphql]: do we need to send authtoken explicitly? that should already be passed as ZM_AUTH_TOKEN cookie

File: src/schema/schema.graphql [hunk 6]
```diff
@@ -280,10 +280,36 @@ type AuthResponse {
 	skin: [Skin]
 }
 
+type ScratchCode{
```
Comment (silentsakky) [inline src/schema/schema.graphql]: space before curly brace

File: src/schema/schema.graphql [hunk 7]
```diff
@@ -280,10 +280,36 @@ type AuthResponse {
...
+	ScratchCode: String
+}
+
+type EnableTwoFactorAuthResponse {
+	secret: [Secret]
+	scratchCodes: [ScratchCode]
```
Comment (silentsakky) [inline src/schema/schema.graphql]: this is mandatory property

File: src/schema/schema.graphql [hunk 8]
```diff
@@ -280,10 +280,36 @@ type AuthResponse {
 	skin: [Skin]
 }
 
+type ScratchCode{
+	ScratchCode: String
```
Comment (silentsakky) [inline src/schema/schema.graphql]: `scratchCode`

File: src/schema/schema.graphql [hunk 9]
```diff
@@ -280,10 +280,36 @@ type AuthResponse {
...
+	secret: [Secret]
+	scratchCodes: [ScratchCode]
+	authToken: [TwoFactorAuthToken]
+}
+
 type ProfileImageChangeResponse {
 	itemId: ID
 }
 
+type TwoFactorAuthToken {
```
Comment (silentsakky) [inline src/schema/schema.graphql]: we already have authtoken type
`type AuthToken`
please reuse it here

File: src/schema/schema.graphql [hunk 10]
```diff
@@ -280,10 +280,36 @@ type AuthResponse {
...
+	_content: String,
+	verifyAccount: Boolean,
+	lifetime: Int
+}
+
+input TwoFactorAuthTokenInput {
```
Comment (silentsakky) [inline src/schema/schema.graphql]: rename to AuthTokenInput

File: src/schema/schema.graphql [hunk 11]
```diff
@@ -280,10 +280,36 @@ type AuthResponse {
...
+	authToken: [TwoFactorAuthToken]
+}
+
 type ProfileImageChangeResponse {
 	itemId: ID
 }
 
+type TwoFactorAuthToken {
+	_content: String,
+	verifyAccount: Boolean,
```
Comment (silentsakky) [inline src/schema/schema.graphql]: not sure if we are going to use these two properties

File: src/schema/schema.graphql [hunk 12]
```diff
@@ -280,10 +280,36 @@ type AuthResponse {
...
+	lifetime: Int
+}
+
+input TwoFactorAuthTokenInput {
+	_content: String,
+	verifyAccount: Boolean,
```
Comment (silentsakky) [inline src/schema/schema.graphql]: same as https://github.com/Zimbra/zm-api-js-client/pull/193/files#r286814042

File: src/schema/schema.graphql [hunk 13]
```diff
@@ -280,10 +280,36 @@ type AuthResponse {
...
 }
 
+type ScratchCode{
+	ScratchCode: String
+}
+
+type EnableTwoFactorAuthResponse {
+	secret: [Secret]
+	scratchCodes: [ScratchCode]
+	authToken: [TwoFactorAuthToken]
```
Comment (silentsakky) [inline src/schema/schema.graphql]: are we using authtoken anywhere? it will already be available in cookie

File: src/batch-client/index.ts [hunk 14]
```diff
@@ -432,6 +433,28 @@ export class ZimbraBatchClient {
...
+					: null,
+				authToken: options.authToken,
+				twoFactorCode: options.twoFactorCode
+					? {
+							_content: options.twoFactorCode
+					  }
+					: null
+			},
+			namespace: Namespace.Account
+		}).then(res => res);
```
Comment (silentsakky) [inline src/batch-client/index.ts]: you don't need then here

File: src/schema/schema.graphql (hunk 7 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: mandatory property cannot be set to a type

File: src/schema/schema.graphql (hunk 10 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: removed it, as it was not needed

File: src/batch-client/index.ts (hunk 14 above)
Comment (panditsavitags) [reply src/batch-client/index.ts]: done

File: src/schema/schema.graphql (hunk 13 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 12 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 11 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 10 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 9 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 8 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 7 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 6 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 5 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/schema/schema.graphql (hunk 4 above)
Comment (panditsavitags) [reply src/schema/schema.graphql]: done

File: src/batch-client/index.ts (hunk 3 above)
Comment (panditsavitags) [reply src/batch-client/index.ts]: done

File: src/batch-client/index.ts (hunk 2 above)
Comment (panditsavitags) [reply src/batch-client/index.ts]: done

File: src/batch-client/index.ts (hunk 1 above)
Comment (panditsavitags) [reply src/batch-client/index.ts]: done

File: src/batch-client/index.ts (hunk 3 above)
Comment (SantoshKShetty) [reply src/batch-client/index.ts]: I think passing data in `_content` is required here because of the structure like `<password>{content}</password>`

**Pull request merged**